src/utils/listener/block_downloader.go

```diff
@@ -226,3 +198,124 @@ func (self *BlockDownloader) run() error {
+			block, err := self.downloadOneBlock(height, lastProcessedBlockHash, peer)
+			if err != nil {
+				block = nil
+			}
+			mtx.Lock()
+			if block != nil {
+				out = append(out, block)
+			}
+			mtx.Unlock()
+			wg.Done()
+		})
+	}
+
+	// Wait for all workers to finish
+	wg.Wait()
+
+	return
+}
+
+func (self *BlockDownloader) downloadOneBlock(height uint64, lastProcessedBlockHash arweave.Base64String, peer string) (block *arweave.Block, err error) {
+	ctx := context.WithValue(self.Ctx, arweave.ContextDisablePeers, true)
+	if len(peer) > 0 {
+		// Force using this peer if set
+		// Otherwise use arweave.net
+		ctx = context.WithValue(ctx, arweave.ContextForcePeer, peer)
+	}
+
+	block, resp, err := self.client.GetBlockByHeight(ctx, int64(height))
+	if err != nil {
+		self.Log.
+			WithError(err).
+			WithField("height", height).
+			WithField("peer", peer).
+			Error("Failed to download block")
+		return
+	}
+
+	if !block.IsValid() {
+		self.Log.
+			WithField("height", height).
+			WithField("peer", peer).
+			WithField("age", resp.Header().Get("Age")).
+			WithField("x-trace", resp.Header().Get("X-Trace")).
+			Error("Block hash isn't valid")
+		self.monitor.GetReport().BlockDownloader.Errors.BlockValidationErrors.Inc()
+		err = errors.New("block isn't valid")
+		return
+	}
+
+	if len(lastProcessedBlockHash) > 0 &&
+		!bytes.Equal(lastProcessedBlockHash, block.PreviousBlock) {
+		self.Log.
+			WithField("height", height).
+			WithField("age", resp.Header().Get("Age")).
+			WithField("x-trace", resp.Header().Get("X-Trace")).
+			WithField("last_processed_block_hash", lastProcessedBlockHash.Base64()).
+			WithField("previous_block", block.PreviousBlock.Base64()).
+			WithField("peer", peer).
+			Warn("Previous block hash isn't valid")
+		// This doesn't mean it's a bad block, we may have a fork
+	}
+
+	return
+}
+
+func (self *BlockDownloader) vote(blocks []*arweave.Block) (out *arweave.Block, err error) {
+	minNumberOfVotes := math.Min((math.Round(2.0 * float64(len(blocks)) / 3.0)), float64(len(blocks)))
+	self.Log.WithField("min_number_of_votes", minNumberOfVotes).Info("Voting for the best block")
+	// Pick the most common block
+	votes := make(map[string]int)
+	for _, block := range blocks {
+		votes[block.IndepHash.Base64()] += 1
+	}
+
+	if len(votes) > 1 {
+		self.Log.
+			WithField("votes", votes).
+			WithField("min_number_of_votes", minNumberOfVotes).
+			Warn("Detected non-unanimous vote")
+	} else {
+		self.Log.WithField("num_votes", len(blocks)).Debug("Unanimous vote")
+	}
+
+	// Pick the most voted block
+	for hash, count := range votes {
+		if float64(count) < minNumberOfVotes {
+			continue
+		}
+
+		// This block got enough votes
+		for _, block := range blocks {
+			if block.IndepHash.Base64() == hash {
+				out = block
+				break
+			}
+		}
+		break
+	}
+
+	if out == nil {
+		self.Log.WithField("votes", votes).
+			WithField("min_number_of_votes", minNumberOfVotes).
+			Error("None of the blocks got enough vote")
```

Comment:
can we also please block hashes here (with their number of votes)?

Reply:
'log block hashes here' :-)

Reply:
votes is a map hash->num votes and it is logged here